cpp/src/io/functions.cpp

```diff
@@ -249,32 +277,35 @@ using namespace cudf::io::detail::parquet;
 namespace detail_parquet = cudf::io::detail::parquet;
 
 // Freeform API wraps the detail reader class API
-table_with_metadata read_parquet(read_parquet_args const& args, rmm::mr::device_memory_resource* mr)
+table_with_metadata read_parquet(parquet_reader_options const& options,
+                                 rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  detail_parquet::reader_options options{
-    args.columns, args.strings_to_categorical, args.use_pandas_metadata, args.timestamp_type};
-  auto reader = make_reader<detail_parquet::reader>(args.source, options, mr);
-
-  if (args.row_groups.size() > 0) {
-    return reader->read_row_groups(args.row_groups);
-  } else if (args.skip_rows != -1 || args.num_rows != -1) {
-    return reader->read_rows(args.skip_rows, args.num_rows);
+  auto reader    = make_reader<detail_parquet::reader>(options.source(), options, mr);
+  auto skip_rows = options.get(parquet_reader_options::size_type_param_id::SKIP_ROWS);
+  auto num_rows  = options.get(parquet_reader_options::size_type_param_id::NUM_ROWS);
+
+  auto row_groups = options.row_groups();
+  if (row_groups.size() > 0) {
+    return reader->read_row_groups(row_groups);
+  } else if ((skip_rows != -1) || (num_rows != -1)) {
+    return reader->read_rows(skip_rows, num_rows);
   } else {
     return reader->read_all();
   }
```

Comment:
The three functions called here all call impl::read. I would love if we could remove these different versions and just forward to impl::read. Can you please explore this option?

Reply:
I was also thinking same, but was thinking about a follow-up PR, but I will try to incorporate if it doesn't bloat the PR.